extensions/arc/deployment/src/main/java/io/quarkus/arc/deployment/ConfigBuildStep.java

```diff
@@ -315,23 +375,108 @@ private static void collectInterfacesRec(ClassInfo current, IndexView index, Set
         }
     }
 
+    @BuildStep
+    void validateConfigMappingsInjectionPoints(
+            ValidationPhaseBuildItem validationPhase,
+            List<ConfigClassBuildItem> configClasses,
+            BuildProducer<ConfigMappingBuildItem> configMappings) {
+
+        Set<ConfigMappingBuildItem> toRegister = new HashSet<>();
+        for (InjectionPointInfo injectionPoint : validationPhase.getContext().getInjectionPoints()) {
+            Type type = injectionPoint.getType();
+            for (ConfigClassBuildItem configClass : configClasses) {
+                if (configClass.isMapping()
+                        && type.name().equals(DotName.createSimple(configClass.getConfigClass().getName()))) {
+
+                    AnnotationTarget target = injectionPoint.getTarget();
+                    AnnotationInstance mapping = null;
+                    if (target.kind().equals(FIELD)) {
+                        mapping = target.asField().annotation(CONFIG_MAPPING_NAME);
+                    } else if (target.kind().equals(METHOD)) {
+                        List<Type> parameters = target.asMethod().parameters();
+                        for (int i = 0; i < parameters.size(); i++) {
+                            Type parameter = parameters.get(i);
+                            if (parameter.name().equals(type.name())) {
+                                Set<AnnotationInstance> parameterAnnotations = getParameterAnnotations(
+                                        validationPhase.getBeanProcessor().getBeanDeployment(), target.asMethod(), i);
+                                mapping = Annotations.find(parameterAnnotations, CONFIG_MAPPING_NAME);
+                            }
+                        }
+                    }
+
+                    AnnotationValue annotationPrefix = null;
+                    if (mapping != null) {
+                        annotationPrefix = mapping.value("prefix");
+                    }
+
+                    String prefix = annotationPrefix != null ? annotationPrefix.asString() : configClass.getPrefix();
+                    configMappings.produce(new ConfigMappingBuildItem(configClass.getConfigClass(), prefix));
+                }
+            }
+        }
+
+        boolean noRemovedBeans = validationPhase.getContext().removedBeans().isEmpty();
+        for (ConfigClassBuildItem configClass : configClasses) {
+            if (configClass.isMapping()
+                    && (noRemovedBeans || configClass.getConfigClass().isAnnotationPresent(Unremovable.class))) {
```

Comment:
I'm not sure I understand the condition here. We produce a `ConfigMappingBuildItem` if `quarkus.arc.remove-unused-beans=false` (because otherwise there will be at least one removed bean in 99.99% cases) or if the config class is annotated with `@Unremovable`? That's probably not correct. I think that we should not produce the build item if the relevant synthetic bean was removed. So in theory, we should be able to match the relevant `BeanInfo`.

Reply:
I guess I was trying to be too smart by only looking for the annotation if no beans were removed.

Reply:
Ah, I added that, because the Config TCK expects a `DeploymentException`. On the TCK we have `quarkus.arc.remove-unused-beans=false`.

Reply:
I still can't understand ;-). Currently, it reads: "register the config mapping if no bean was removed OR the mapping is annotated with `@Unremovable`". 

You should probably inject the `io.quarkus.arc.deployment.ArcConfig` in the build step and test `ArcConfig.shouldEnableBeanRemoval()` intead of `validationPhase.getContext().removedBeans().isEmpty()`. 

Also even a mapping bean that is not annotated with `@Unremovable` may be considered "unremovable", i.e. when it's injected somewhere. Such a config mapping should be registered as well, or?